src/libpmemfile/preload.c

```diff
@@ -0,0 +1,2024 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * preload.c - The main code controlling the preloadable version of pmemfile. To
+ * understand the code start from the routine pmemfile_preload_constructor() -
+ * this should run before the application starts, and while there is only a
+ * single thread of execution in the process. To understand the syscall
+ * routing logic look at the routine hook(...), this is called by the libc
+ * syscall intercepting code every time libc would issue a syscall instruction.
+ * The hook() routine decides if a syscall should be handled by kernel, or
+ * by pmemfile ( and which pmemfile pool ).
+ */
+
+#define _GNU_SOURCE
+
+#include <assert.h>
+#include <fcntl.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <syscall.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <string.h>
+#include <errno.h>
+#include <setjmp.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <limits.h>
+
+#include <asm-generic/errno.h>
+
+#include "libsyscall_intercept_hook_point.h"
+#include "libpmemfile-posix.h"
+#include "util.h"
+
+/*
+ * Some syscalls that are missing on older kernels.
+ */
+#ifndef SYS_renameat2
+#define SYS_renameat2 316
+#endif
+
+#ifndef SYS_execveat
+#define SYS_execveat 322
+#endif
+
+#ifndef SYS_copy_file_range
+#define SYS_copy_file_range 326
+#endif
+
+#ifndef SYS_preadv2
+#define SYS_preadv2 327
+#endif
+
+#ifndef SYS_pwritev2
+#define SYS_pwritev2 328
+#endif
+
+static pf_noreturn void
+exit_group_no_intercept(int ret)
+{
+	syscall_no_intercept(SYS_exit_group, ret);
+	__builtin_unreachable();
+}
+
+static inline void
+FATAL(const char *str)
+{
+	syscall_no_intercept(SYS_write, 2, str, strlen(str));
+	exit_group_no_intercept(128 + 7);
+}
+
+#include "sys_util.h"
+
+#include "fd_pool.h"
+
+#include "preload.h"
+
+static int hook(long syscall_number,
+			long arg0, long arg1,
+			long arg2, long arg3,
+			long arg4, long arg5,
+			long *result);
+
+static bool syscall_number_filter[0x200];
+static bool syscall_needs_fd_rlock[0x200];
+static bool syscall_needs_fd_wlock[0x200];
+static bool syscall_needs_pmem_cwd_rlock[0x200];
+static bool syscall_has_fd_first_arg[0x200];
+
+static pthread_rwlock_t fd_lock = PTHREAD_RWLOCK_INITIALIZER;
+
+static struct pool_description pools[0x100];
+static int pool_count;
+
+/*
+ * A separate place to keep track of fds used to hold mount points open, in
+ * the previous array. The application should not be aware of these, thus
+ * whenever these file descriptors are encountered during interposing, -EBADF
+ * must be returned. The contents of this array does not change after startup.
+ */
+static bool mount_point_fds[PMEMFILE_MAX_FD + 1];
+
+static struct fd_association fd_table[PMEMFILE_MAX_FD + 1];
+
+static pthread_rwlock_t pmem_cwd_lock = PTHREAD_RWLOCK_INITIALIZER;
```

Comment:
Theoretically you could wrap both the mutex and cwd in a struct like this:

```
struct protected_resource {
  pthread_mutex_t lock;
  void *__resource; 
} pres;

```
and have a macro like:
```
struct foo *res = PROTECTED_RESOURCE_BEGIN(struct foo, &pres);
...
PROTECTED_RESOURCE_END(&pres);
```

This makes sense in higher level languages where RAII is a thing and you could associate the resource with a mutex by having a locked smart pointer.

In C, and in the case of your particular code, I would make write locks follow the same pattern as read locks. And if it wouldn't be particularly painful, maybe pass in the cwd variable as an argument to those handler functions (if required, otherwise NULL), that way you would have to have a lock if you wanted the variable.